File: internal/models/pack.go
// agency-gateway/internal/models/pack.go
package models

import "fmt"

// PackAgent represents a single agent entry in a pack team.
type PackAgent struct {
	Name      string   `yaml:"name" validate:"required"`
	Preset    string   `yaml:"preset" validate:"required"`
	Workspace *string  `yaml:"workspace"`
	Role      string   `yaml:"role" validate:"omitempty,oneof=standard coordinator function" default:"standard"`
	AgentType *string  `yaml:"agent_type"`
	Host      *string  `yaml:"host"`
	Skills    []string `yaml:"skills"`
	Connectors []string `yaml:"connectors"`
}

// PackChannel represents a communication channel defined in a pack.
type PackChannel struct {
	Name    string `yaml:"name" validate:"required"`
	Topic   string `yaml:"topic"`
	Private bool   `yaml:"private"`
}

// PackCredential represents a credential required by a pack.
type PackCredential struct {
	Name        string `yaml:"name" validate:"required"`
	Description string `yaml:"description"`
	Required    bool   `yaml:"required" default:"true"`
}

// PackMissionAssignment binds a mission to an agent within a pack.
type PackMissionAssignment struct {
	Mission string `yaml:"mission" validate:"required"`
	Agent   string `yaml:"agent" validate:"required"`
}

// PackRequires lists the dependencies of a pack.
type PackRequires struct {
	Connectors []string `yaml:"connectors"`
	Presets    []string `yaml:"presets"`
	Services   []string `yaml:"services"`
	Skills     []string `yaml:"skills"`
	Workspaces []string `yaml:"workspaces"`
	Policies   []string `yaml:"policies"`
}

// PackTeam defines the team composition within a pack.
type PackTeam struct {
	Name     string        `yaml:"name" validate:"required"`
	Agents   []PackAgent   `yaml:"agents" validate:"required"`
	Channels []PackChannel `yaml:"channels"`
}

// Validate implements cross-field validation for PackTeam.
// Checks that agents is non-empty and that agent/channel names are unique.
func (pt *PackTeam) Validate() error {
	if len(pt.Agents) == 0 {
		return fmt.Errorf("pack must define at least one agent")
	}

	agentSeen := make(map[string]bool)
	for _, a := range pt.Agents {
		if agentSeen[a.Name] {
			return fmt.Errorf("duplicate agent names: %s", a.Name)
		}
		agentSeen[a.Name] = true
	}

	channelSeen := make(map[string]bool)
	for _, c := range pt.Channels {
		if channelSeen[c.Name] {
			return fmt.Errorf("duplicate channel names: %s", c.Name)
		}
		channelSeen[c.Name] = true
	}

	return nil
}

// PackConfig is the schema for pack.yaml files.
type PackConfig struct {
	Kind                 string           `yaml:"kind" validate:"required,oneof=pack" default:"pack"`
	Name                 string           `yaml:"name" validate:"required"`
	Version              string           `yaml:"version" default:"1.0.0"`
	Description          string           `yaml:"description"`
	Author               string           `yaml:"author"`
	License              string           `yaml:"license,omitempty"`
	Requires             PackRequires     `yaml:"requires"`
	Team                 PackTeam         `yaml:"team"`
	Credentials          []PackCredential `yaml:"credentials"`
	Policy               map[string]interface{} `yaml:"policy"`
	RecommendedConnectors []string        `yaml:"recommended_connectors"`
	MissionAssignments   []PackMissionAssignment `yaml:"mission_assignments"`
}

// Validate implements cross-field validation for PackConfig.
// Applies per-agent defaults and delegates to PackTeam for structural checks.
func (pc *PackConfig) Validate() error {
	for i := range pc.Team.Agents {
		if pc.Team.Agents[i].Role == "" {
			pc.Team.Agents[i].Role = "standard"
		}
	}
	return pc.Team.Validate()
}

File: images/models/pack.py
"""Pydantic models for pack schema — declarative team composition."""

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
from typing import Optional, Literal


class PackAgent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    preset: str
    workspace: Optional[str] = None
    role: Literal["standard", "coordinator", "function"] = "standard"
    agent_type: Optional[str] = None
    host: Optional[str] = None
    skills: list[str] = Field(default_factory=list)
    connectors: list[str] = Field(default_factory=list)


class PackChannel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    topic: str = ""
    private: bool = False


class PackCredential(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    description: str = ""
    required: bool = True


class PackMissionAssignment(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mission: str
    agent: str


class PackRequires(BaseModel):
    model_config = ConfigDict(extra="forbid")

    connectors: list[str] = Field(default_factory=list)
    presets: list[str] = Field(default_factory=list)
    services: list[str] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    workspaces: list[str] = Field(default_factory=list)
    policies: list[str] = Field(default_factory=list)


class PackTeam(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    agents: list[PackAgent]
    channels: list[PackChannel] = Field(default_factory=list)

    @field_validator("agents")
    @classmethod
    def agents_not_empty(cls, v: list[PackAgent]) -> list[PackAgent]:
        if not v:
            raise ValueError("Pack must define at least one agent")
        return v

    @model_validator(mode="after")
    def no_duplicate_names(self) -> "PackTeam":
        agent_names = [a.name for a in self.agents]
        if len(agent_names) != len(set(agent_names)):
            dupes = [n for n in agent_names if agent_names.count(n) > 1]
            raise ValueError(f"Duplicate agent names: {set(dupes)}")
        channel_names = [c.name for c in self.channels]
        if len(channel_names) != len(set(channel_names)):
            dupes = [n for n in channel_names if channel_names.count(n) > 1]
            raise ValueError(f"Duplicate channel names: {set(dupes)}")
        return self


class PackConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["pack"] = "pack"
    name: str
    version: str = "1.0.0"
    description: str = ""
    author: str = ""
    license: str = ""
    requires: PackRequires = Field(default_factory=PackRequires)
    team: PackTeam
    credentials: list[PackCredential] = Field(default_factory=list)
    policy: Optional[dict] = None
    recommended_connectors: list[str] = Field(default_factory=list)
    mission_assignments: list[PackMissionAssignment] = Field(default_factory=list)
